File: TP4Python/TP4.py
#!/usr/bin/python3

import sys, os
COUT_UNIFORME = 0
PROFONDEUR_LIMITE = 1
APPROF_ITERATIF = 2
A_ETOILE = 3
IDA_ETOILE = 4
MAX = 1024
SIZE = 3

class Noeud:
    def __init__(self, taquin, poids, heuristique, profondeur, precedent, suivant, suite):
        self.taquin = taquin
        self.poids = poids
        self.heuristique  = heuristique
        self.profondeur = profondeur
        self.precedent = precedent
        self.suivant = suivant
        self.suite = suite

class Liste:
    def __init__(self, noeud):
        self.noeud = noeud

def est_but(taquin):
    for i in range(0,(SIZE * SIZE) - 1):
        if(taquin[i] + 1 != taquin[i+1]):
            return False
    return True

def est_solvable(taquin):
    count=0
    for i in range(0,(SIZE * SIZE) - 1):
        for j in range(i+1,(SIZE*SIZE)):
            if(taquin[i] != 0 and taquin[j] != 0):
                if(taquin[i] > taquin[j]):
                    count += 1
    return count%2 == 0

def print_taquin(taquin, indice_fin):
    indice=(SIZE*SIZE)
    if(indice_fin < indice):
        indice=indice_fin
    for i in range(0,indice):
        print("%d       ",taquin[i])
        if((i+1)%3 == 0):
            print("\n")
    print("\n")

def print_taquin_ligne(taquin, indice_fin):
    indice=(SIZE*SIZE)
    if(indice_fin < indice):
        indice=indice_fin
    for i in range(0,indice):
        print("%d ",taquin[i])

def get_position_item(taquin, item):
    for i in range(0,SIZE*SIZE):
        if(taquin[i] == item):
            return i
    return -1

def deplacementGauche(taquin, position):
    if(position % SIZE != 0):
        taquin[position] = taquin[position - 1]
        taquin[position - 1] = 0
        return 1
    else:
        return 0

def deplacementDroite(taquin, position):
    if(position % SIZE != SIZE - 1):
        taquin[position] = taquin[position + 1]
        taquin[position + 1] = 0
        return 1
    else:
        return 0

def deplacementHaut(taquin, position):
    if(position >= SIZE):
        taquin[position] = taquin[position - SIZE]
        taquin[position - SIZE] = 0
        return 1
    else:
        return 0

def deplacementBas(taquin, position):
    if(position < (SIZE * (SIZE - 1))):
        taquin[position] = taquin[position + SIZE]
        taquin[position + SIZE] = 0
        return 1
    else:
        return 0

def est_solution(taquin, sol):
    size = len(sol)
    position_blanc = 0
    for i in range(0,size):
        position_blanc = get_position_item(taquin, 0)
        if(sol[i].upper == 'G'):
            deplacementGauche(taquin,position_blanc)
        elif(sol[i].upper == 'D'):
            deplacementDroite(taquin,position_blanc)
        elif(sol[i].upper == 'H'):
            deplacementHaut(taquin,position_blanc)
        elif(sol[i].upper == 'B'):
            deplacementBas(taquin,position_blanc)
    return est_but(taquin)

def successeurs(taquin, taquins):
    position_blanc = get_position_item(taquin, 0)
    suite = ""

    taquin_gauche = []
    taquin_droite = []
    taquin_haut = []
    taquin_bas = []

    taquin_gauche = taquin.copy()
    taquin_droite = taquin.copy()
    taquin_haut = taquin.copy()
    taquin_bas = taquin.copy()

    if(deplacementGauche(taquin_gauche,position_blanc)):
        taquins.append(taquin_gauche)
        suite += "G"
    if(deplacementDroite(taquin_droite,position_blanc)):
        taquins.append(taquin_droite)
        suite += "D"
    if(deplacementHaut(taquin_haut,position_blanc)):
        taquins.append(taquin_haut)
        suite += "H"
    if(deplacementBas(taquin_bas,position_blanc)):
        taquins.append(taquin_bas)
        suite += "B"
    return suite

def swap(x, y):
    temp = x
    x = y
    y = temp

def write_taquin(data, taquin, indice_fin):
    indice=(SIZE*SIZE)-1
    if(indice_fin < indice):
        indice=indice_fin
    for i in range(0,indice+1):
        data.write(str(taquin[i]))

def permutation(taquin, debut, fin, data):
    if(debut == fin):
        print_taquin(taquin, fin)
        write_taquin(data, taquin, fin)
        data.write('\n')
    else:
        for i in range(debut, fin + 1):
            swap(taquin[debut], taquin[i])
            permutation(taquin, debut+1, fin, data)
            swap(taquin[debut], taquin[i])

def empiler(pile, taquin, poids, suite, profondeur, heuristique):
    if ((not pile)):
        return False
    noeud = Noeud(taquin,poids,heuristique,profondeur,False,pile.noeud,suite[:])
    if(pile.noeud):
        pile.noeud.precedent = noeud
    pile.noeud = noeud

def depiler(pile):
    if (not pile):
        return False

    noeudDepile = pile.noeud

    if (pile and pile.noeud):
        pile.noeud = noeudDepile.suivant
    return noeudDepile

def retirerUniforme(pile):
    if (not pile):
        return False
    actuel = pile.noeud
    petit=actuel

    while (actuel):
        if(actuel.poids < petit.poids):
            petit = actuel
        actuel = actuel.suivant
    if(petit):
        if(petit==pile.noeud):
            pile.noeud = petit.suivant
            if(pile.noeud):
                pile.noeud.precedent=False
        else:
            petit.precedent.suivant = petit.suivant
        if(petit.suivant):
            petit.suivant.precedent = petit.precedent
    return petit

def retirerEtoile(pile):
    if (not pile):
        return False
    actuel = pile.noeud
    petit=actuel

    while (actuel):
        if((actuel.poids + actuel.heuristique) < (petit.poids + petit.heuristique)):
            petit = actuel
        actuel = actuel.suivant
    if(petit):
        if(petit==pile.noeud):
            pile.noeud = petit.suivant
            if(pile.noeud):
                pile.noeud.precedent = False
        else:
            petit.precedent.suivant = petit.suivant
        if(petit.suivant):
            petit.suivant.precedent = petit.precedent
    return petit

def afficherListe(liste):
    if (not liste):
        return False
    actuel = liste.noeud

    while (actuel):
        print_taquin(actuel.taquin,(SIZE*SIZE))
        actuel = actuel.suivant

def listeVide(liste):
    return not liste.noeud

def taquin_similaire(taquin1, taquin2, size):
    for i in range(0, size):
        if(taquin1[i] != taquin2[i]):
            return 0
    return 1

def a_ete_traite(liste, taquin):
    if (not liste):
        return 0
    actuel = liste.noeud

    while (actuel):
        if(taquin_similaire(actuel.taquin,taquin,(SIZE*SIZE))):
            return 1
        actuel=actuel.suivant
    return 0

def inserer(pile, taquin, poids, suite, strategie, profondeur, heuristique):
    empiler(pile,taquin,poids,suite,profondeur, heuristique)

def retirer(pile, strategie):
    if(strategie == COUT_UNIFORME):
        return retirerUniforme(pile)
    elif((strategie == IDA_ETOILE) or (strategie == A_ETOILE)):
        return retirerEtoile(pile)
    elif(strategie == PROFONDEUR_LIMITE or strategie == APPROF_ITERATIF):
        return depiler(pile)
    return False

def ask_depth():
    depth = input("Jusqu'à quelle profondeur ?\n")
    return depth

def nbCasesF(taquinInit, taquinBut):
    j=0
    for i in range(0, SIZE*SIZE):
        if(taquinInit[i] != taquinBut[i]):
            j+=1
    return j

def manhattan(taquinInit, taquinBut):
    sumPos=0
    for i in range(0,SIZE*SIZE):
        if(taquinInit[i] != 0):
            posInit = get_position_item(taquinInit, taquinInit[i])
            initX = posInit//SIZE
            initY = posInit%SIZE
            posBut = get_position_item(taquinBut, taquinInit[i])
            butX = posBut//SIZE
            butY = posBut%SIZE
            sumPos += abs(butX - initX) + abs(butY - initY)
    return sumPos

def recherche(taquin, taquinBut, strategie, dataWriter):
    profondeur_maximale = False
    profondeur=0
    if(not est_solvable(taquin)):
        dataWriter.write(" -1\n")
        return False
    l = Liste(False)
    l_done = Liste(False)
    inserer(l,taquin,0,"",strategie,0, manhattan(taquin,taquinBut))
    
    if(strategie == PROFONDEUR_LIMITE or strategie == APPROF_ITERATIF):
        profondeur_maximale=ask_depth()
    if(strategie == PROFONDEUR_LIMITE):
        profondeur=profondeur_maximale

    while(not listeVide(l)):
        e = retirer(l,strategie)
        inserer(l_done,e.taquin,e.poids,e.suite,strategie,e.profondeur,e.heuristique)
        if(est_but(e.taquin)):
            dataWriter.write(" " + str(len(e.suite)) + " ")
            dataWriter.write(e.suite)
            dataWriter.write("\n")
            return
        else:
            if((strategie != PROFONDEUR_LIMITE and strategie != APPROF_ITERATIF) or (e.profondeur < profondeur)):
                taquins = []
                suiteSucc = successeurs(e.taquin,taquins)
                for i in range(0, len(suiteSucc)):
                    if((not a_ete_traite(l_done,taquins[i])) and (not a_ete_traite(l,taquins[i]))):
                        suite = e.suite[:]
                        suite += suiteSucc[i]
                        inserer(l,taquins[i],e.poids+1,suite,strategie,e.profondeur+1, manhattan(taquins[i],taquinBut))
        if (listeVide(l) and strategie == APPROF_ITERATIF and profondeur < profondeur_maximale):
            inserer(l,taquin,0,"",strategie,0, manhattan(taquin,taquinBut))
            l_done = Liste(False)
            profondeur+=1

def recherche_idaEtoile(seuil, taquin, taquinBut, dataWriter):
    if(not est_solvable(taquin)):
        dataWriter.write(" -1\n")
        return False
    l = Liste(False)
    l_done = Liste(False)
    inserer(l,taquin,0,"",IDA_ETOILE,0, manhattan(taquin,taquinBut))
    min=4096


    while (not listeVide(l)):
        e = retirer(l,IDA_ETOILE)
        inserer(l_done,e.taquin,e.poids,e.suite,IDA_ETOILE,e.profondeur,e.heuristique)
        if(est_but(e.taquin)):
            dataWriter.write(" " + str(len(e.suite)) + " ")
            dataWriter.write(e.suite)
            dataWriter.write("\n")
            return False
        else:
            taquins = []
            suiteSucc = successeurs(e.taquin,taquins)
            for i in range(0,len(suiteSucc)):
                if((not a_ete_traite(l_done,taquins[i])) and (not a_ete_traite(l,taquins[i]))):
                    suite = e.suite[:]
                    suite += suiteSucc[i]
                    fonctionFFils = manhattan(taquins[i],taquinBut)
                    feprime = e.poids + 1 + fonctionFFils
                    if(feprime <= seuil):
                        inserer(l,taquins[i],e.poids+1,suite,IDA_ETOILE,e.profondeur+1, fonctionFFils)
                    elif(feprime < min):
                        min = feprime
    return recherche_idaEtoile(min, taquin, taquinBut, dataWriter)

def Main():
    if(len(sys.argv) != 2):
        print("Invalid number of parameters")
        return False
    print("Waiting...\n")
    script_dir = os.path.dirname(__file__) # absolute directory the file is in
    rel_path = "../E_O/sortie.txt"
    abs_file_path = os.path.join(script_dir, rel_path)
    fdL = open(sys.argv[1], "r")
    fdW = open(abs_file_path, "w")
    taquinsListen = fdL.read().split("\n")
    taquinBut = list(map(int, list(taquinsListen[0])))
    for i in range(1,len(taquinsListen)):
        if(len(taquinsListen[i]) == 9):
            taquinNormal = list(map(int, list(taquinsListen[i])))
            for t in taquinNormal:
                fdW.write(str(t))
            # recherche(taquinNormal, taquinBut, A_ETOILE, fdW)
            recherche_idaEtoile(manhattan(taquinNormal,taquinBut),taquinNormal, taquinBut, fdW)
    fdW.close()
    fdL.close()

if __name__ == '__main__':
    Main()

File: TP4/TP4.c
#include "TP1_2_3.c"
#define COUT_UNIFORME 0
#define PROFONDEUR_LIMITE 1
#define APPROF_ITERATIF 2
#define A_ETOILE 3
#define IDA_ETOILE 4
#define NB_CASES_F 1
#define MANHATTAN 2
#define MANHATTAN_DERIVE 3
#define MAX 200

typedef struct Noeud Noeud;
struct Noeud {
    int *taquin;
    int poids;
    int heuristique;
    int profondeur;
    Noeud *precedent;
    Noeud *suivant;
    char *suite;
};

typedef struct _liste {
    Noeud *noeud;
} Liste;

void empiler(Liste *pile, int *taquin, int poids, char *suite, int profondeur, int heuristique) {
    Noeud *noeud = malloc(sizeof(Noeud));
    if (pile == NULL || noeud == NULL) exit(EXIT_FAILURE);

    noeud->suite = malloc(MAX*sizeof(char));
    strcpy(noeud->suite, suite);
    noeud->taquin = malloc((SIZE * SIZE) * sizeof(int));
    clone_taquin(taquin, noeud->taquin);
    noeud->poids = poids;
    noeud->profondeur = profondeur;
    noeud->heuristique = heuristique;
    noeud->suivant = pile->noeud;
    noeud->precedent = NULL;
    if(pile->noeud) pile->noeud->precedent = noeud;
    pile->noeud = noeud;
    // free(taquin);
}

Noeud *depiler(Liste *pile) {
    if (pile == NULL) exit(EXIT_FAILURE);

    Noeud *noeudDepile = pile->noeud;

    if (pile != NULL && pile->noeud != NULL) pile->noeud = noeudDepile->suivant;
    return noeudDepile;
}

Noeud *retirerUniforme(Liste *pile){
    if (pile == NULL) exit(EXIT_FAILURE);
    Noeud *actuel = pile->noeud;
    Noeud *petit=actuel;

    while (actuel != NULL) {
        if(actuel->poids < petit->poids) petit = actuel;
        actuel = actuel->suivant;
    }
    if(petit != NULL) {
        if(petit==pile->noeud) {
            pile->noeud = petit->suivant;
            if(pile->noeud) pile->noeud->precedent=NULL;
        } else petit->precedent->suivant = petit->suivant;
        if(petit->suivant) {
            petit->suivant->precedent = petit->precedent;
        }
    }
    return petit;
}

Noeud *retirerEtoile(Liste *pile){
    if (pile == NULL) exit(EXIT_FAILURE);
    Noeud *actuel = pile->noeud;
    Noeud *petit=actuel;
    Noeud *new=malloc(sizeof(Noeud));

    while (actuel != NULL) {
        if((actuel->poids + actuel->heuristique) < (petit->poids + petit->heuristique)) petit = actuel;
        actuel = actuel->suivant;
    }
    if(petit != NULL) {
        if(petit==pile->noeud) {
            pile->noeud = petit->suivant;
            if(pile->noeud) pile->noeud->precedent=NULL;
        } else petit->precedent->suivant = petit->suivant;
        if(petit->suivant) {
            petit->suivant->precedent = petit->precedent;
        }
        new->suite = malloc(MAX*sizeof(char));
        strcpy(new->suite, petit->suite);
        new->taquin = malloc((SIZE * SIZE) * sizeof(int));
        clone_taquin(petit->taquin, new->taquin);
        new->poids = petit->poids;
        new->profondeur = petit->profondeur;
        new->heuristique = petit->heuristique;
        new->suivant = NULL;
        new->precedent = NULL;
        free(petit->suite);
        free(petit->taquin);
        free(petit);
    }
    return new;
}

void afficherListe(Liste *liste) {
    if (liste == NULL) exit(EXIT_FAILURE);
    Noeud *actuel = liste->noeud;

    while (actuel != NULL) {
        print_taquin(actuel->taquin,(SIZE*SIZE));
        actuel = actuel->suivant;
    }
}

int listeVide(Liste *liste) {
    return !liste->noeud;
}

int taquin_similaire(int *taquin1, int *taquin2, int size) {
    for(int i=0;i<size;i++) {
        if(taquin1[i] != taquin2[i]) return 0;
    }
    return 1;
}

int a_ete_traite(Liste *liste, int *taquin) {
    if (liste == NULL) return 0;
    Noeud *actuel = liste->noeud;

    while (actuel != NULL) {
        if(taquin_similaire(actuel->taquin,taquin,(SIZE*SIZE))) return 1;
        actuel=actuel->suivant;
    }
    return 0;
}

void inserer(Liste *pile, int *taquin, int poids, char *suite, int strategie, int profondeur, int heuristique) {
    empiler(pile,taquin,poids,suite,profondeur, heuristique);
}

Noeud *retirer(Liste *pile, int strategie) {
    if(strategie == COUT_UNIFORME) return retirerUniforme(pile);
    else if((strategie == IDA_ETOILE) || (strategie == A_ETOILE)) return retirerEtoile(pile);
    else if(strategie == PROFONDEUR_LIMITE || strategie == APPROF_ITERATIF) return depiler(pile);
    return NULL;
}

int ask_depth() {
    int depth;
    printf("Jusqu'à quelle profondeur ?\n");
    scanf("%d",&depth);
    return depth;
}

void viderListe(Liste *liste) {
    if (liste == NULL) return;
    Noeud *actuel=liste->noeud;
    if (actuel == NULL) return;
    while (actuel->suivant != NULL) {
        actuel = actuel->suivant;
        free(actuel->precedent->suite);
        free(actuel->precedent->taquin);
        free(actuel->precedent);
    }
    free(actuel->suite);
    free(actuel->taquin);
    free(actuel);
}

int nbCasesF(int *taquinInit, int *taquinBut) {
    int j=0;
    for(int i=0; i<(SIZE*SIZE); i++) {
        if(taquinInit[i] != taquinBut[i]) j++;
    }
    return j;
}

int manhattan(int *taquinInit, int *taquinBut) {
    int sumPos=0;
    for(int i=0; i<SIZE*SIZE; i++) {
        if(taquinInit[i] != 0) {
            int posInit = get_position_item(taquinInit, taquinInit[i]);
            int initX = posInit/SIZE;
            int initY = posInit%SIZE;
            int posBut = get_position_item(taquinBut, taquinInit[i]);
            int butX = posBut/SIZE;
            int butY = posBut%SIZE;
            sumPos += abs(butX - initX) + abs(butY - initY);
        }
    }
    return sumPos;
}

int deriveManhattan(int *taquinInit, int *taquinBut) {
    int sumPos=0;
    for(int i=0; i<SIZE*SIZE; i++) {
        if(taquinInit[i] != 0) {
            int posInit = get_position_item(taquinInit, taquinInit[i]);
            int initX = posInit/SIZE;
            int initY = posInit%SIZE;
            int posBut = get_position_item(taquinBut, taquinInit[i]);
            int butX = posBut/SIZE;
            int butY = posBut%SIZE;
            sumPos += abs(butX - initX) + abs(butY - initY);
            if(get_position_item(taquinInit, taquinInit[i]) > get_position_item(taquinBut, taquinInit[i])) sumPos +=2;
        }
    }
    return sumPos;
}

void recherche(int *taquin, int *taquinBut, int strategie, dataBufferWriter *dataWriter, int typeHeuristic) {
    int profondeur_maximale;
    int profondeur=0;
    if(!est_solvable(taquin)) {
        write_datas(dataWriter, " -1\n");
        return;
    }
    Liste *l = malloc(sizeof(Liste) * 5000);
    l->noeud = NULL;
    Liste *l_done = malloc(sizeof(Liste) * 5000);
    l_done->noeud = NULL;
    char lasuite[MAX];
    strcpy(lasuite,"");
    if(typeHeuristic == MANHATTAN) inserer(l,taquin,0,lasuite,strategie,0, manhattan(taquin,taquinBut));
    else if(typeHeuristic == MANHATTAN_DERIVE) inserer(l,taquin,0,lasuite,strategie,0, deriveManhattan(taquin,taquinBut));
    else if(typeHeuristic == NB_CASES_F) inserer(l,taquin,0,lasuite,strategie,0, nbCasesF(taquin,taquinBut));
    
    if(strategie == PROFONDEUR_LIMITE || strategie == APPROF_ITERATIF) profondeur_maximale=ask_depth();
    if(strategie == PROFONDEUR_LIMITE) profondeur=profondeur_maximale;

    while(!listeVide(l)) {
        Noeud *e = retirer(l,strategie);
        if(est_but(e->taquin)) {
            char str[5];
            sprintf(str, " %d ", (int)strlen(e->suite));
            write_datas(dataWriter, str);
            write_datas(dataWriter, e->suite);
            write_datas(dataWriter, "\n");
            free(e->suite);
            free(e->taquin);
            free(e);
            viderListe(l);
            viderListe(l_done);
            free(l);
            free(l_done);
            return;
        }
        inserer(l_done,e->taquin,e->poids,e->suite,strategie,e->profondeur,e->heuristique);
        if((strategie != PROFONDEUR_LIMITE && strategie != APPROF_ITERATIF) || (e->profondeur < profondeur)) {
            char suiteSucc[5];
            int **taquins = malloc(4 * sizeof(int *));
            strcpy(suiteSucc,"");
            int row = successeurs(e->taquin,suiteSucc,taquins);
            for(int i=0; i<row; i++) {
                if(!a_ete_traite(l_done,taquins[i]) && !a_ete_traite(l,taquins[i])) {
                    char suite[MAX];
                    strcpy(suite,e->suite);
                    strncat(suite,&suiteSucc[i],1);
                    if(typeHeuristic == MANHATTAN) inserer(l,taquins[i],e->poids+1,suite,strategie,e->profondeur+1, manhattan(taquins[i],taquinBut));
                    else if(typeHeuristic == MANHATTAN_DERIVE) inserer(l,taquins[i],e->poids+1,suite,strategie,e->profondeur+1, deriveManhattan(taquins[i],taquinBut));
                    else if(typeHeuristic == NB_CASES_F) inserer(l,taquins[i],e->poids+1,suite,strategie,e->profondeur+1, nbCasesF(taquins[i],taquinBut));
                }
                free(taquins[i]);
            }
            free(taquins);
            free(e->suite);
            free(e->taquin);
            free(e);
        }
        if (listeVide(l) && strategie == APPROF_ITERATIF && profondeur < profondeur_maximale) {
            char suite[MAX];
            strcpy(suite,"");
            if(typeHeuristic == MANHATTAN) inserer(l,taquin,0,suite,strategie,0, manhattan(taquin,taquinBut));
            else if(typeHeuristic == MANHATTAN_DERIVE) inserer(l,taquin,0,suite,strategie,0, deriveManhattan(taquin,taquinBut));
            else if(typeHeuristic == NB_CASES_F) inserer(l,taquin,0,suite,strategie,0, nbCasesF(taquin,taquinBut));
            free(l_done);
            l_done = malloc(sizeof(Liste) * 5000);
            l_done->noeud = NULL;
            profondeur++;
        }
    }
    viderListe(l);
    viderListe(l_done);
    free(l);
    free(l_done);
}

void recherche_idaEtoile(int seuil, int *taquin, int *taquinBut, dataBufferWriter *dataWriter, int typeHeuristic) {
    if(!est_solvable(taquin)) {
        write_datas(dataWriter, " -1\n");
        return;
    }
    Liste *l = malloc(sizeof(Liste) * 5000);
    l->noeud = NULL;
    Liste *l_done = malloc(sizeof(Liste) * 5000);
    l_done->noeud = NULL;
    char lasuite[MAX];
    strcpy(lasuite,"");
    if(typeHeuristic == MANHATTAN) inserer(l,taquin,0,lasuite,IDA_ETOILE,0, manhattan(taquin,taquinBut));
    else if(typeHeuristic == MANHATTAN_DERIVE) inserer(l,taquin,0,lasuite,IDA_ETOILE,0, deriveManhattan(taquin,taquinBut));
    else if(typeHeuristic == NB_CASES_F) inserer(l,taquin,0,lasuite,IDA_ETOILE,0, nbCasesF(taquin,taquinBut));
    int min=4096;

    while (!listeVide(l)) {
        Noeud *e = retirer(l,IDA_ETOILE);
        if(est_but(e->taquin)) {
            char str[5];
            sprintf(str, " %d ", (int)strlen(e->suite));
            write_datas(dataWriter, str);
            write_datas(dataWriter, e->suite);
            write_datas(dataWriter, "\n");
            free(e->suite);
            free(e->taquin);
            free(e);
            viderListe(l);
            viderListe(l_done);
            free(l);
            free(l_done);
            return;
        }
        inserer(l_done,e->taquin,e->poids,e->suite,IDA_ETOILE,e->profondeur,e->heuristique);
        char suiteSucc[5];
        int **taquins = malloc(4 * sizeof(int *));
        strcpy(suiteSucc,"");
        int row = successeurs(e->taquin,suiteSucc,taquins);
        for(int i=0; i<row; i++) {
            if(!a_ete_traite(l_done,taquins[i]) && !a_ete_traite(l,taquins[i])) {
                char suite[MAX];
                strcpy(suite,e->suite);
                strncat(suite,&suiteSucc[i],1);
                int fonctionFFils = 0;
                if(typeHeuristic == MANHATTAN) fonctionFFils = manhattan(taquins[i],taquinBut);
                else if(typeHeuristic == MANHATTAN_DERIVE) fonctionFFils = deriveManhattan(taquins[i],taquinBut);
                else if(typeHeuristic == NB_CASES_F) fonctionFFils = nbCasesF(taquins[i],taquinBut);
                int feprime = e->poids + 1 + fonctionFFils;
                if(feprime <= seuil) inserer(l,taquins[i],e->poids+1,suite,IDA_ETOILE,e->profondeur+1, fonctionFFils);
                else if(feprime < min) min = feprime;
            }
            free(taquins[i]);
        }
        free(taquins);
        free(e->suite);
        free(e->taquin);
        free(e);
    }
    viderListe(l);
    viderListe(l_done);
    free(l);
    free(l_done);
    return recherche_idaEtoile(min, taquin, taquinBut, dataWriter, typeHeuristic);
}

int main(int argc, char *argv[]) {
    char in[MAX] = "";
    printf("Input file ?\n");
    scanf("%s",in);
    char out[MAX] = "";
    printf("Output file ?\n");
    scanf("%s",out);
    dataBufferReader *dataReader = createBufferReader(in);
    dataBufferWriter *dataWriter = createBufferWriter(out);
    int typeAnalysis = -1;
    while(typeAnalysis != 1 && typeAnalysis != 2) {
        printf("Analysis Type ?\nType 1 for A*\t-- or --\tType 2 for IDA*\n");
        scanf("%d",&typeAnalysis);
    }
    int typeHeuristic = -1;
    while(typeHeuristic != 1 && typeHeuristic != 2 && typeHeuristic != 3) {
        printf("Heuristic Type ?\nType 1 for nbCasesF\t-- or --\tType 2 for Manhattan\t-- or --\tType 3 for Manhattan drift\n");
        scanf("%d",&typeHeuristic);
    }
    printf("Waiting...\n");
    int *sBut = malloc((SIZE * SIZE) * sizeof(int));
    int isSet = 0;
    char s[SIZE*SIZE];
    strcpy(s,"");
    char c;
    while((c = getCurrentChar(dataReader)) != -1) {
        if(c == '\n') {
            if(!isSet) {
                isSet = 1;
                for(int i=0; i<(SIZE*SIZE); i++) sBut[i] = s[i] - '0';
            } else {
                int *taquin = malloc((SIZE * SIZE) * sizeof(int));
                for(int i=0; i<(SIZE*SIZE); i++) taquin[i] = s[i] - '0';
                write_taquin(dataWriter, taquin, SIZE*SIZE);
                if(typeAnalysis == 1 && typeHeuristic == 1) recherche(taquin, sBut, A_ETOILE, dataWriter, NB_CASES_F);
                else if(typeAnalysis == 1 && typeHeuristic == 2) recherche(taquin, sBut, A_ETOILE, dataWriter, MANHATTAN);
                else if(typeAnalysis == 1 && typeHeuristic == 3) recherche(taquin, sBut, A_ETOILE, dataWriter, MANHATTAN_DERIVE);
                else if(typeAnalysis == 2 && typeHeuristic == 1) recherche_idaEtoile(nbCasesF(taquin,sBut),taquin, sBut, dataWriter, NB_CASES_F);
                else if(typeAnalysis == 2 && typeHeuristic == 2) recherche_idaEtoile(manhattan(taquin,sBut),taquin, sBut, dataWriter, MANHATTAN);
                else if(typeAnalysis == 2 && typeHeuristic == 3) recherche_idaEtoile(manhattan(taquin,sBut),taquin, sBut, dataWriter, MANHATTAN_DERIVE);
                free(taquin);
            }
            memset(s,0,SIZE*SIZE);
        } else strncat(s,&c,1);
        consumeChar(dataReader);
    }
    free(sBut);
    destroyBufferWriter(dataWriter);
    destroyBufferReader(dataReader);
}